flocker/control/test/test_persistence.py

```diff
@@ -249,3 +303,88 @@ def test_complex_keys(self):
         self.assertEqual(node_state, wire_decode(wire_encode(node_state)))
+
+
+class FakeMigration(object):
+    """
+    A simple fake migration class, used to test ``migrate_configuration``.
+    """
+    @classmethod
+    def configuration_v1_v2(cls, config):
+        config = json.loads(config)
+        if config['version'] != 1:
+            raise ConfigurationMigrationError(
+                "Supplied configuration was not a valid v1 config."
+            )
+        return json.dumps({"version": 2, "configuration": "fake"})
+
+    @classmethod
+    def configuration_v2_v3(cls, config):
+        config = json.loads(config)
+        if config['version'] != 2:
+            raise ConfigurationMigrationError(
+                "Supplied configuration was not a valid v2 config."
+            )
+        return json.dumps({"version": 3, "configuration": "fake"})
+
+
+class ConfigurationMigrationTests(SynchronousTestCase):
+    """
+    Tests for ``ConfigurationMigration`` and ``migrate_configuration``.
+    """
+    v1_config = json.dumps({"version": 1})
+
+    def test_error_on_undefined_migration_path(self):
+        """
+        A ``ConfigurationMigrationError`` is raised if a migration path
+        from one version to another cannot be found in the supplied
+        migration class.
+        """
+        e = self.assertRaises(
+            ConfigurationMigrationError,
+            migrate_configuration, 1, 4, self.v1_config, FakeMigration
+        )
+        expected_error = (
+            u'Unable to find a migration path for a version 1 '
+            u'to version 4 configuration. No migration method exists '
+            u'for v3 to v4.'
+        )
+        self.assertEqual(e.message, expected_error)
+
+    def test_sequential_migrations(self):
+        """
+        A migration from one configuration version to another will
+        sequentially perform all necessary upgrades, e.g. v1 to v2 followed
+        by v2 to v3.
+        """
+        v2_config = migrate_configuration(1, 2, self.v1_config, FakeMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, FakeMigration)
+        self.assertEqual(result, FakeMigration.configuration_v2_v3(v2_config))
+
+    def test_v1_latest_configuration(self):
+        """
+        A V1 JSON configuration blob is transformed to the latest
+        configuration, with the result validating when loaded.
+        """
+        upgraded_json = migrate_configuration(
+            1, _CONFIG_VERSION, V1_TEST_DEPLOYMENT_JSON,
+            ConfigurationMigration)
+        upgraded_config = wire_decode(upgraded_json)
+        expected_configuration = Configuration(
+            version=_CONFIG_VERSION, deployment=TEST_DEPLOYMENT
+        )
+        self.assertEqual(upgraded_config, expected_configuration)
+
+    def test_v1_v2_configuration(self):
```

Comment:
I think this test belongs somewhere else.  One sensible collection of tests is for `migrate_configuration`, independent of the specifics our actual configuration schema/types/values/objects.  Another sensible collection is for the upgrades defined by `ConfigurationMigration`.